### ISSUE TITLE
Doesn’t calculate layout when message text edit

### ISSUE THREAD

**Issue by gellertk** [ekazaev/ChatLayout#34]
Hi can you fix this case, for example when message has one line text, and I edit this message to to two lined text, it doesn’t make space between edited message and last message

**Comment by ekazaev** [comment]
Hi @gellertk 

Sorry. I didnt really understand your request. Can you please explain in details and attach the screenshots or some code if possible if possible? If you reload the cell with the message in my understanding in should expand it.

**Comment by gellertk** [comment]
Sorry from me) i just need possibility for recalculate layout if item size changed

https://user-images.githubusercontent.com/54902273/192499224-4929c13d-e0ec-4de9-8ea5-dcd3a87968c6.mov

**Comment by ekazaev** [comment]
@gellertk 

You can try it yourself to see that such behaviour works correctly. Modify this function in `DefaultChatController` to try yourself.

```swift
    func deleteMessage(with id: UUID) {
        let lastMessage = messages.removeLast()
        guard case var .text(string) = lastMessage.data else {
            return
        }
        string += " \(string)"
        messages.append(RawMessage(id: lastMessage.id, date: lastMessage.date, data: .text(string), userId: lastMessage.userId, status: lastMessage.status))
        repopulateMessages()
    }
```

![changing_text_demo](https://user-images.githubusercontent.com/20243258/192761264-6fda9ae2-2842-42d4-9718-e6b6191d7421.gif)

**Comment by ekazaev** [comment]
@gellertk Please let me know if you are satisfied with the answer

**Comment by gellertk** [comment]
We use DifferenceKit for collection updating, and after some research i found - in example app after collectionView.reload( calling 
```swift
public override func invalidationContext(forPreferredLayoutAttributes preferredAttributes: UICollectionViewLayoutAttributes, withOriginalAttributes originalAttributes: UICollectionViewLayoutAttributes) -> UICollectionViewLayoutInvalidationContext {
```
But in my case it never called
And we also use sizeForItem and returning .exact sizes, do you have any thoughts about it?

**Comment by ekazaev** [comment]
@gellertk I think it might be because you are using.exact sizes tbh. Can you check that when collection reloads this cell it asks delegate for the new size and you provide the updated one? It also of course can be an issue on my side. I don’t know anyone who uses exact sizes as the performance is poor in this case.

**Comment by gellertk** [comment]
if i understand right u talking about sizeForItem function call after collection reloading?
No it doesnt call, cause doesnt call func invalidationContext, and func invalidationContext should call sizeForItem
```swift
public override func invalidationContext(forPreferredLayoutAttributes preferredAttributes: UICollectionViewLayoutAttributes, withOriginalAttributes originalAttributes: UICollectionViewLayoutAttributes) -> UICollectionViewLayoutInvalidationContext {
...
let newItemSize = itemSize(with: preferredMessageAttributes) //line 508
```

**Comment by ekazaev** [comment]
@gellertk Let me check that scenario and Ill come back to you

**Comment by ekazaev** [comment]
@gellertk I tested your case in the example app and I can confirm that the layout works correctly. 
You can try it yourself. In addition to the code snippet I gave you above - change text cell size calculation in the `sizeForItem` in `DefaultChatCollectionDataSource` to :
```swift
                    let size = CGSize(width: chatLayout.layoutFrame.width, height: 18 + 18.0 * CGFloat(Int(string.count / 30)))
                    return .exact(size)
```

And it works as expected: 
![changing_text_demo](https://user-images.githubusercontent.com/20243258/193011896-0bfdca99-fc6d-4a54-96d0-12c4574c3e95.gif)

Please, make sure that you are calling `reloadItem` for the cell you are updating text.

Thats is all the information I can provide you by now, I will check few more things - but most likely the issue is in your code.

**Comment by ekazaev** [comment]
@gellertk I did other checks and can confirm that the layout works as expected. Once again, you must call `reloadItem` so that layout knows that it needs to remeasure the cell. No matter if you do it manually or using the `DifferenceKit` or any other differencing algorithm. 

The easiest way to debug the fact that you called a reload method is to add
```swift
        print("\(#function) \(updateItems)")
```

to 

```swift
    /// Notifies the layout object that the contents of the collection view are about to change.
    public override func prepare(forCollectionViewUpdates updateItems: [UICollectionViewUpdateItem]) {
```

in `ChatLayout`

That will reset `calculatedSize` in the layout model so `ChatLayout` will try to remeasure your cell and will call the delegate for an updated size.

Basically, every time you want to update the cell you should expect to see something like the next line in the log:
```
prepare(forCollectionViewUpdates:) [R(0,167)]
```
with an index path of a cell that you are updating.

**Comment by ekazaev** [comment]
@gellertk You can even use available in ios 15 and onward `reconfigureItems` method. Add to the patch I gave you next code at line 477 in `ChatViewController`.

```swift
            if changeSet.count == 1, changeSet[0].elementUpdated.count == 1 {
                let element = changeSet[0].elementUpdated[0]
                dataSource.sections[element.section].cells[element.element] = sections[element.section].cells[element.element]
                if #available(iOS 15.0, *) {
                    self.collectionView.reconfigureItems(at: [IndexPath(item: element.element, section: element.section)])
                    completion?()
                }
                return
            }
```

I am afraid that it is maximum I can do to help you. 

You need to look for the problem in your code. You will actually have to share it or change the Example app the way so I could check your problem.

Best of luck.

**Comment by ilia3546** [comment]
@ekazaev Hi there! 
I made a simple app which reproduces this problem — https://github.com/ilia3546/ChatLayout-SizeBugSample
Please check it, thank you ;)

**Comment by ekazaev** [comment]
Hi @ilia3546 @gellertk 

I checked your example. Thank you.

So the problem there is in the is in the cell implementation. Collection view asks each cell for its preferred size, which doesn't change in your case. Label just thrown there without any constraints and collection uses those values to decide if it should change something in the layout or not. You just changed label inside but for collection view nothing has changed.

You can see in the Reveal that the cell stayed the same: 
![Uploading Screenshot 2022-10-04 at 12.51.07.png…]()

You need to notify the collection that the preferred attributes has changed so that collection would start asking the layout about what and how it should change. 

In your case if you don't want to rely on auto layout calculation you need to add to `SampleCollectionCell` something like:

```
    override func preferredLayoutAttributesFitting(_ layoutAttributes: UICollectionViewLayoutAttributes) -> UICollectionViewLayoutAttributes {
        layoutAttributes.size = label.bounds.size
        return layoutAttributes
    }
```
 so that collection view will start to ask layout how it should and if it should invalidate the cell. As preferred attributes didn't change, collection does not ask layout if it should invalidate the cell and layout doesn't ask delegate about new size. As you are able to calculate the size programmatically - you may want to move your calculations to the `preferredLayoutAttributesFitting` function instead of using the delegate for the same purpose.

Of course it is my interpretation of how `UICollectionView` works as it is not documented and the layout is implemented accordingly to my understanding. I may review my understanding in the future but currently it is so.

**Comment by ekazaev** [comment]
@ilia3546 @gellertk 

I played around with it and it seems there is no harm to update those values from the delegate on reload call. So you'll get the expansion without overriding default `preferredLayoutAttributesFitting` implementation. Keep in mind that yet `UICollectionView` doesn't ask its layout if it should invalidate the cell so it may lead to some visual inconsistencies as some compensations may not be applied. But you'll get the desired expansion out of the box.

Please check the version `1.2.11` and let me know if you are satisfied.

**Comment by ilia3546** [comment]
Awesome! Thank you so much!

**Comment by ekazaev** [comment]
@ilia3546 My pleasure. Keep in mind that `preferredLayoutAttributesFitting` is actually the right place for such manipulations.
